### ISSUE TITLE
hyper: pausing request body is signaling end-of-stream

### ISSUE THREAD

**Issue by seanmonstar** [curl/curl#8600]
### I did this

When pausing the sending of a request body, curl's body data callback returns `HYPER_POLL_READY` with a `null` chunk, meaning the body is finished. This is likely the cause of the failures of tests 670-673.

### I expected the following

Instead, as long the body might continue in the future, curl should save a waker and return `HYPER_POLL_PENDING`. When the user code asks curl to resume again, curl internally can "wake" the waker, telling hyper to poll for data another time.

To inline the [docs](https://docs.rs/hyper/latest/hyper/ffi/fn.hyper_body_set_data_func.html):

> If there is more data to send, but it isn’t yet available, a `hyper_waker` should be saved from the `hyper_context *` argument, and `HYPER_POLL_PENDING` should be returned. You must wake the saved waker to signal the task when data is available.

It looks like the code in `c-hyper.c` already has support for this when waiting for expect-100 vs a timer, so it would just be expanding that.

### curl/libcurl version

```
curl 7.82.1-DEV (x86_64-pc-linux-gnu) libcurl/7.82.1-DEV OpenSSL/1.0.2g zlib/1.2.8 librtmp/2.3 Hyper/0.14.17 OpenLDAP/2.4.42
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP TrackMemory UnixSockets
```

### operating system

Linux grunt 5.10.13-x86_64-linode141 #1 SMP Thu Feb 4 13:56:42 EST 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by bagder** [comment]
The `lib/c-hyper.c:hyper_body_chunk()` function currently makes no attempt to understand when a pause was requested, and it not knowing about it makes it instead handle it as end of data.

**Comment by tivrfoa** [comment]
I'm trying to run the test 670 with `./runtests.pl -f -g 670`, but I get:<br>
*There was no content at all in the file **log/server.input.***

Do I need to pass any other parameter to test it?
How to create this file, please?


```sh
./runtests.pl -f -g 670
********* System characteristics ******** 
* curl 7.83.0-DEV (x86_64-pc-linux-gnu) 
* libcurl/7.83.0-DEV OpenSSL/1.1.1f zlib/1.2.11 Hyper/0.14.17
* Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP TrackMemory UnixSockets
* Disabled: 
* Host: lesco-Lenovo-IdeaPad-S145-15API
* System: Linux lesco-Lenovo-IdeaPad-S145-15API 5.13.0-35-generic #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
* OS: linux
*
*** DISABLES memory tracking when using threaded resolver
*
* Servers: HTTP-IPv6 HTTP-unix FTP-IPv6 
* Env: 
* Seed: 254767
***************************************** 
Warning: test670 is explicitly disabled
test 0670...[Request pause from mime read callback: multi]
GNU gdb (Ubuntu 10.2-0ubuntu1~20.04~1) 10.2
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./libtest/lib670...
Argument list to give program being debugged when it is started is "http://127.0.0.1:34247/670".
(gdb) r
Starting program: /home/lesco/dev/github/curl/tests/libtest/lib670 http://127.0.0.1:34247/670
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
URL: http://127.0.0.1:34247/670
* STATE: INIT => CONNECT handle 0x55555565fa48; line 1834 (connection #-5000)
* Added connection 0. The cache now contains 1 members
* family0 == v4, family1 == v6
*   Trying 127.0.0.1:34247...
* STATE: CONNECT => CONNECTING handle 0x55555565fa48; line 1895 (connection #0)
* Connected to 127.0.0.1 (127.0.0.1) port 34247 (#0)
* STATE: CONNECTING => PROTOCONNECT handle 0x55555565fa48; line 2027 (connection #0)
* STATE: PROTOCONNECT => DO handle 0x55555565fa48; line 2050 (connection #0)
* Time for the Hyper dance
> POST /670 HTTP/1.1
> Host: 127.0.0.1:34247
> Accept: */*
> Content-Length: 142
> Content-Type: multipart/form-data; boundary=------------------------3433323135333231
> 
* STATE: DO => DID handle 0x55555565fa48; line 2146 (connection #0)
* STATE: DID => PERFORMING handle 0x55555565fa48; line 2265 (connection #0)
* Hyper: [1] user body write aborted: early end, expected 49 more bytes
* multi_done: status: 56 prem: 1 done: 0
* The cache now contains 0 members
* Closing connection 0
* Expire cleared (transfer 0x55555565fa48)
[Inferior 1 (process 38159) exited normally]
(gdb) quit

 670: protocol FAILED!
 There was no content at all in the file log/server.input.
 Server glitch? Total curl failure? Returned: 0

 - abort tests
TESTDONE: 1 tests were considered during 306 seconds.
TESTDONE: 0 tests out of 1 reported OK: 0%

TESTFAIL: These test cases failed: 670
```

**Comment by tivrfoa** [comment]
>
> A little trick can be to do a separate build without hyper and run to get to see how it looks and works when everything works correctly.

Thanks a lot! Good idea.

I configured without hyper and the test `670` also failed after `make test`:
```sh
TESTDONE: 1480 tests were considered during 385 seconds.
TESTDONE: 1309 tests out of 1314 reported OK: 99%

TESTFAIL: These test cases failed: 670 671 672 673 1540 

make[1]: *** [Makefile:807: quiet-test] Error 1
make[1]: Leaving directory '/home/lesco/dev/github/curl-no-hyper/tests'
make: *** [Makefile:1669: test] Error 2
```

Running only this test case:

```sh
curl-no-hyper/tests (master)$ ./runtests.pl 670
********* System characteristics ******** 
* curl 7.83.0-DEV (x86_64-pc-linux-gnu) 
* libcurl/7.83.0-DEV OpenSSL/1.1.1f zlib/1.2.11
* Features: alt-svc AsynchDNS Debug HSTS HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP TrackMemory UnixSockets
* Disabled: 
* Host: lesco-Lenovo-IdeaPad-S145-15API
* System: Linux lesco-Lenovo-IdeaPad-S145-15API 5.13.0-35-generic #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
* OS: linux
*
*** DISABLES memory tracking when using threaded resolver
*
* Servers: HTTP-IPv6 HTTP-unix FTP-IPv6 
* Env: 
* Seed: 241556
***************************************** 
test 0670...[Request pause from mime read callback: multi]

 670: data FAILED:
--- log/check-expected	2022-03-23 19:46:24.084985518 -0300
+++ log/check-generated	2022-03-23 19:46:24.084985518 -0300
@@ -1,7 +1,21 @@
+Testing url http://127.0.0.1:33981/670[LF]
+curl_easy_pause[LF]
+part->kind = 4[LF]
+part->kind = 3[LF]
+newstate & KEEP_RECV_PAUSE =        0x0[LF]
+get this handle going again[LF]
+curl_easy_pause[LF]
+No change. state = 0[LF]
 HTTP/1.1 200 OK[CR][LF]
+curl_easy_pause[LF]
+No change. state = 0[LF]
 Date: Tue, 09 Nov 2010 14:49:00 GMT[CR][LF]
 Server: test-server/fake swsclose[CR][LF]
 Connection: close[CR][LF]
+curl_easy_pause[LF]
+No change. state = 0[LF]
 Content-Type: text/html[CR][LF]
 [CR][LF]
 hello[LF]
+curl_easy_pause[LF]
+No change. state = 0[LF]

 - abort tests
TESTDONE: 1 tests were considered during 3 seconds.
TESTDONE: 0 tests out of 1 reported OK: 0%

TESTFAIL: These test cases failed: 670
```

**Comment by tivrfoa** [comment]
ops. The test works.<br>

The `printf`s that I put were causing the test to fail ...

```sh
tests (master)$ ./runtests.pl 670
********* System characteristics ******** 
* curl 7.83.0-DEV (x86_64-pc-linux-gnu) 
* libcurl/7.83.0-DEV OpenSSL/1.1.1f zlib/1.2.11
* Features: alt-svc AsynchDNS Debug HSTS HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP TrackMemory UnixSockets
* Disabled: 
* Host: lesco-Lenovo-IdeaPad-S145-15API
* System: Linux lesco-Lenovo-IdeaPad-S145-15API 5.13.0-35-generic #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
* OS: linux
*
*** DISABLES memory tracking when using threaded resolver
*
* Servers: HTTP-IPv6 HTTP-unix FTP-IPv6 
* Env: 
* Seed: 241556
***************************************** 
test 0670...[Request pause from mime read callback: multi]
--pd---e--- OK (1   out of 1  , remaining: 00:00, took 2.686s, duration: 00:02)
TESTDONE: 1 tests were considered during 3 seconds.
TESTDONE: 1 tests out of 1 reported OK: 100%
```

**Comment by tivrfoa** [comment]
Is this code correct?
I'm asking because without Hyper the test prints `* We are completely uploaded and fine`

Is Hyper replacing that part?
I just found it a bit confusing because even with Hyper support the `else` can still be executed, right?

transfer.c

```c
#ifdef USE_HYPER
  if(conn->datastream) {
    result = conn->datastream(data, conn, &didwhat, done, select_res);
    if(result || *done)
      return result;
  }
  else {
#endif
  /* We go ahead and do a read if we have a readable socket or if
     the stream was rewound (in which case we have data in a
     buffer) */
  if((k->keepon & KEEP_RECV) && (select_res & CURL_CSELECT_IN)) {
    result = readwrite_data(data, conn, k, &didwhat, done, comeback);
    if(result || *done)
      return result;
  }

  /* If we still have writing to do, we check if we have a writable socket. */
  if((k->keepon & KEEP_SEND) && (select_res & CURL_CSELECT_OUT)) {
    /* write */

    result = readwrite_upload(data, conn, &didwhat);
    if(result)
      return result;
  }
#ifdef USE_HYPER
  }
#endif
```

**Comment by bagder** [comment]
> I'm asking because without Hyper the test prints "* We are completely uploaded and fine" Is Hyper replacing that part?

It is, yes. The is output by the `readwrite_upload()` function.

> I just found it a bit confusing because even with Hyper support the else can still be executed, right?

Yes it can. Like for non-HTTP protocols libcurl is built to support, when `conn->datastream` is not set and thus still a NULL pointer.

**Comment by tivrfoa** [comment]
> The `lib/c-hyper.c:hyper_body_chunk()` function currently makes no attempt to understand when a pause was requested, and it not knowing about it makes it instead handle it as end of data.

I set a breakpoint in this function and apparently it's not even called.
But I might be doing something wrong again ...

```sh
(gdb) info b
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   0x00005555555ec8e7 in hyper_body_chunk at c-hyper.c:176
(gdb) r
Starting program: /home/lesco/dev/github/curl/tests/libtest/lib670 http://127.0.0.1:35409/670
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
URL: http://127.0.0.1:35409/670
Testing url http://127.0.0.1:35409/670
* STATE: INIT => CONNECT handle 0x55555565fe58; line 1834 (connection #-5000)
* Added connection 0. The cache now contains 1 members
* family0 == v4, family1 == v6
*   Trying 127.0.0.1:35409...
* STATE: CONNECT => CONNECTING handle 0x55555565fe58; line 1895 (connection #0)
* Connected to 127.0.0.1 (127.0.0.1) port 35409 (#0)
* STATE: CONNECTING => PROTOCONNECT handle 0x55555565fe58; line 2027 (connection #0)
* STATE: PROTOCONNECT => DO handle 0x55555565fe58; line 2050 (connection #0)
* Time for the Hyper dance
> POST /670 HTTP/1.1
> Host: 127.0.0.1:35409
> Accept: */*
> Content-Length: 142
> Content-Type: multipart/form-data; boundary=------------------------3433323135333231
> 
* STATE: DO => DID handle 0x55555565fe58; line 2146 (connection #0)
* STATE: DID => PERFORMING handle 0x55555565fe58; line 2265 (connection #0)
* Hyper: [1] user body write aborted: early end, expected 49 more bytes
* multi_done: status: 56 prem: 1 done: 0
* The cache now contains 0 members
* Closing connection 0
* Expire cleared (transfer 0x55555565fe58)
[Inferior 1 (process 4495) exited normally]
```
![2022-03-24 21-57-22-curl-hyper-HYPER_TASK_EMPTY](https://user-images.githubusercontent.com/30683/160034291-1c96e168-02ff-446a-abd3-a6f133fe5465.png)

**Comment by tivrfoa** [comment]
> When pausing the sending of a request body, curl's body data callback returns `HYPER_POLL_READY` with a `null` chunk, meaning the body is finished. This is likely the cause of the failures of tests 670-673.

The change below solves 670 and 673, but break many things. =)

```diff
-------------------------------- lib/c-hyper.c --------------------------------
index b795a47fe..0f7b72361 100644
@@ -683,9 +683,11 @@ static int uploadstreamed(void *userdata, hyper_context *ctx,
     data->state.hresult = result;
     return HYPER_POLL_ERROR;
   }
-  if(!fillcount)
+  if(!fillcount) {
     /* done! */
     *chunk = NULL;
+    return HYPER_POLL_PENDING;
+  }
   else {
     hyper_buf *copy = hyper_buf_copy((uint8_t *)data->state.ulbuf, fillcount);
     if(copy)

```

So maybe some condition to check if it can return `HYPER_POLL_PENDING`?
I have no idea how curl or hyper work, I'm just trying to help =)

**Comment by bagder** [comment]
> I have no idea how curl or hyper work, I'm just trying to help =)

Proposing a patch that "breaks many things" isn't helping much ...

**Comment by tivrfoa** [comment]
> > I have no idea how curl or hyper work, I'm just trying to help =)
> 
> Proposing a patch that "breaks many things" isn't helping much ...

Obviously I am not proposing that patch

**Comment by bagder** [comment]
Nobody is working on this issue and it is mentioned as [a known problem with the hyper build](https://github.com/curl/curl/wiki/Hyper), Feel free to reopen when there's work or anything else happening that might change this.

**Comment by nnethercote** [comment]
@seanmonstar fixed this issue in #9070 in June 2022, AFAICT. I think this means that the "pausing transfers" item can be removed from the "Outstanding work" list on [the wiki](https://github.com/curl/curl/wiki/Hyper)?

**Comment by nnethercote** [comment]
> @seanmonstar fixed this issue in #9070 in June 2022, AFAICT. I think this means that the "pausing transfers" item can be removed from the "Outstanding work" list on [the wiki](https://github.com/curl/curl/wiki/Hyper)?

I have removed it.
